core/common/src/DateTimePeriod.kt

```diff
@@ -8,21 +8,69 @@ package kotlinx.datetime
 import kotlinx.datetime.internal.*
 import kotlinx.datetime.serializers.DatePeriodIso8601Serializer
 import kotlinx.datetime.serializers.DateTimePeriodIso8601Serializer
+import kotlinx.datetime.serializers.DatePeriodComponentSerializer
+import kotlinx.datetime.serializers.DateTimePeriodComponentSerializer
 import kotlin.math.*
 import kotlin.time.Duration
 import kotlinx.serialization.Serializable
 
 /**
  * A difference between two [instants][Instant], decomposed into date and time components.
  *
- * The date components are: [years], [months], [days].
+ * The date components are: [years] ([DateTimeUnit.YEAR]), [months] ([DateTimeUnit.MONTH]), [days] ([DateTimeUnit.DAY]).
  *
- * The time components are: [hours], [minutes], [seconds], [nanoseconds].
+ * The time components are: [hours] ([DateTimeUnit.HOUR]), [minutes] ([DateTimeUnit.MINUTE]),
+ * [seconds] ([DateTimeUnit.SECOND]), [nanoseconds] ([DateTimeUnit.NANOSECOND]).
  *
- * A `DateTimePeriod` can be constructed using the same-named constructor function,
- * [parsed][DateTimePeriod.parse] from a string, or returned as the result of instant arithmetic operations (see [Instant.periodUntil]).
- * All these functions can return a [DatePeriod] value, which is a subtype of `DateTimePeriod`,
- * a special case that only stores date components, if all time components of the result happen to be zero.
+ * The time components are not independent and always overflow into one another.
+ * Likewise, months overflow into years.
+ * For example, there is no difference between `DateTimePeriod(months = 24, hours = 2, minutes = 63)` and
+ * `DateTimePeriod(years = 2, hours = 3, minutes = 3)`.
+ *
+ * All components can also be negative: for example, `DateTimePeriod(months = -5, days = 6, hours = -3)`.
+ * Whereas `months = 5` means "5 months after," `months = -5` means "5 months earlier."
+ *
+ * Since, semantically, a [DateTimePeriod] is a combination of [DateTimeUnit] values, in cases when the period is a
+ * fixed time interval (like "yearly" or "quarterly"), please consider using [DateTimeUnit] directly instead:
+ * for example, instead of `DateTimePeriod(months = 6)`, one could use `DateTimeUnit.MONTH * 6`.
```

Comment:
What is the justification for this recommendation? Can we document it here?

Reply:
> What is the justification for this recommendation?

The justification is that it's the right tool for the job, and the benefits are numerous.
* There are more operations available on `DateTimeUnit`. If you have a thing like `DateTimePeriod(months = 6)`, to find out how many such periods fit between two instants, you have to write a loop that iteratively adds the period; with `DateTimeUnit.MONTH * 6`, you use `Instant.until`. If you want to add a period several times, you can use `plus(n, unit)` as a simple and performant operation, but there's no `plus(n, period)`, and in general, there's seemingly no way to implement it in general other than with a loop.
* Automatically inferred types for `DateTimeUnit` are more flexible: `val period = when (input) { DAILY -> DateTimeUnit.DAY; MONTHLY -> DateTimeUnit.MONTH }` is automatically a `DateTimeUnit.DateBased`, allowing you to pass it to `LocalDate`-based operations; `val period = DateTimeUnit.HOUR` can automatically be used with `Instant` values without specifying a time zone; if you go the `DateTimePeriod` way, you have to remember to use `DatePeriod` (or try casting to it), or use a `Duration` instead when your periods are time-based, etc. Why do that if `DateTimeUnit` works out of the box?
* For some typical use cases (like periodic reminders, etc.), the limitations of `DateTimeUnit` (that it can't be negative or zero) are actually useful: you won't accidentally fire an alarm every `DateTimePeriod(hours = -2)`.

> Can we document it here?

I'm not sure how to fit this justification into the text naturally.

> Yes, I also do not agree with this recommendation.

Could you elaborate?

Reply:
> I'm not sure how to fit this justification into the text naturally.

I would just cite some of the example you've given here. Eg: 

```kt
/**
 * Since, semantically, a [DateTimePeriod] is a combination of [DateTimeUnit] values, in cases when the period is a
 * fixed time interval (like "yearly" or "quarterly"), please consider using [DateTimeUnit] directly instead. 
 * [DateTimeUnit] allows more idiomatic datetime arithmetic than an equivalent [DateTimePeriod] with operations
 * such as [Instant.plus]. An example of a [DateTimePeriod] which is functionally equivalent to a [DateTimeUnit] is
 * `DateTimePeriod(months = 6)` and `DateTimeUnit.MONTH * 6`. 
 * /
```